File: 05_fp_challenges/01_product_except_self/Solution01.hs
-- Product of Array Except Self  (LeetCode 238)
-- =============================================
-- Given `nums`, return `out` where out[i] = product of every element EXCEPT nums[i].
-- No division. O(n).
--   productExceptSelf [1,2,3,4] == [24,12,8,6]
--   productExceptSelf [-1,1,0,-3,3] == [0,0,9,0,0]

module Main where

import Data.ByteString (isSuffixOf)

iteration :: [Int] -> [Int] -> [Int]
iteration _ [] = []
iteration (p : pref) (s : suf) = (p * s) : iteration pref suf

productExceptSelf :: [Int] -> [Int]
productExceptSelf nums = iteration pref (tail suf)
  where
    pref = scanl (*) 1 nums
    suf = reverse (scanl (*) 1 (reverse nums))

-- ---- tests ----
main :: IO ()
main = do
  check (productExceptSelf [1, 2, 3, 4]) [24, 12, 8, 6]
  check (productExceptSelf [-1, 1, 0, -3, 3]) [0, 0, 9, 0, 0]
  check (productExceptSelf [2, 3]) [3, 2]
  check (productExceptSelf [5]) [1]
  putStrLn "all passed"
  where
    check got want
      | got == want = putStrLn ("  PASS  " ++ show got)
      | otherwise = error ("  FAIL  got " ++ show got ++ " want " ++ show want)

File: 05_fp_challenges/01_product_except_self/Solution01.java
// Product of Array Except Self  (LeetCode 238)
// =============================================
// Given nums, return out where out[i] = product of every element EXCEPT nums[i].
// No division. O(n).
//   productExceptSelf([1,2,3,4]) == [24,12,8,6]
//   productExceptSelf([-1,1,0,-3,3]) == [0,0,9,0,0]

import java.util.Arrays;

public class Solution01 {

    static int[] productExceptSelf(int[] nums) {
        throw new UnsupportedOperationException("TODO");
    }

    // ---- tests ----
    public static void main(String[] args) {
        check(productExceptSelf(new int[]{1,2,3,4}),     new int[]{24,12,8,6});
        check(productExceptSelf(new int[]{-1,1,0,-3,3}), new int[]{0,0,9,0,0});
        check(productExceptSelf(new int[]{2,3}),         new int[]{3,2});
        check(productExceptSelf(new int[]{5}),           new int[]{1});
        System.out.println("all passed");
    }

    static void check(int[] got, int[] want) {
        if (!Arrays.equals(got, want))
            throw new AssertionError("FAIL got " + Arrays.toString(got) + " want " + Arrays.toString(want));
        System.out.println("  PASS  " + Arrays.toString(got));
    }
}
